packages/store/addon/-private/identifiers/cache.ts

```diff
@@ -0,0 +1,426 @@
+  configuredForgetMethod = method;
+}
+
+export function setIdentifierResetMethod(method: ResetMethod): void {
+  configuredResetMethod = method;
+}
+
+function defaultGenerationMethod(data: ResourceIdentifierObject, bucket: string): string {
+  if (isNonEmptyString(data.lid)) {
+    return data.lid;
+  }
+  let { type, id } = data;
+  if (isNonEmptyString(id)) {
+    return `@ember-data:lid-${normalizeModelName(type)}-${id}`;
+  }
+  return uuidv4();
+}
+
+const IdentifierCaches = new WeakMap<object, IdentifierCache>();
+
+export function identifierCacheFor(store: object): IdentifierCache {
+  let cache = IdentifierCaches.get(store);
+
+  if (cache === undefined) {
+    cache = new IdentifierCache();
+    IdentifierCaches.set(store, cache);
+  }
+
+  return cache;
+}
+
+function defaultEmptyCallback(...args: any[]): any {}
+
+let DEBUG_MAP;
+if (DEBUG) {
+  DEBUG_MAP = new WeakMap<StableRecordIdentifier, StableRecordIdentifier>();
+}
+
+export class IdentifierCache {
+  // Typescript still leaks private properties in the final
+  // compiled class, so we may want to move these from _underscore
+  // to a WeakMap to avoid leaking
+  private _cache = {
+    lids: Object.create(null) as IdentifierMap,
+    types: Object.create(null) as TypeMap,
+  };
+  private _generate: GenerationMethod;
+  private _update: UpdateMethod;
+  private _forget: ForgetMethod;
+  private _reset: ResetMethod;
+
+  constructor() {
+    // we cache the user configuredGenerationMethod at init because it must
+    // be configured prior and is not allowed to be changed
+    this._generate = configuredGenerationMethod || defaultGenerationMethod;
+    this._update = configuredUpdateMethod || defaultEmptyCallback;
+    this._forget = configuredForgetMethod || defaultEmptyCallback;
+    this._reset = configuredResetMethod || defaultEmptyCallback;
+  }
+
+  // allows us to peek without generating when needed
+  // useful for the "create" case when we need to see if
+  // we are accidentally overwritting something
+  peekRecordIdentifier(resource: ResourceIdentifierObject, shouldGenerate: true): StableRecordIdentifier;
+  peekRecordIdentifier(resource: ResourceIdentifierObject, shouldGenerate: false): StableRecordIdentifier | undefined;
+  peekRecordIdentifier(
+    resource: ResourceIdentifierObject,
+    shouldGenerate: boolean = false
+  ): StableRecordIdentifier | undefined {
+    // short circuit if we're already the stable version
+    if (isStableIdentifier(resource)) {
+      if (DEBUG) {
+        // TODO should we instead just treat this case as a new generation skipping the short circuit?
+        if (!(resource.lid in this._cache.lids) || this._cache.lids[resource.lid] !== resource) {
+          throw new Error(`The supplied identifier ${resource} does not belong to this store instance`);
+        }
+      }
+      return resource;
+    }
+
+    // `type` must always be present
+    if (DEBUG) {
+      if (!isNonEmptyString(resource.type)) {
```

Comment:
Should we use assert for this instead of throw new Eerror?

Reply:
throwing vs assert has advantages for

- typescript tooling
- test harness
- debugging

Reply:
Should we stop using `Ember.assert` completely and switch? cc @mike-north @rwjblue